### ISSUE TITLE
 No gradients provided for any variable, check your graph for ops that do not support gradients

### ISSUE THREAD

**Issue by fangpin** [zsdonghao/tensorlayer#137]
I am learning GANs. I want to keep some variable not trained during my training epochs. So I just use minimize function and pass network.all_params to parameter var_list.
But I get  No gradients provided for any variable, check your graph for ops that do not support gradients.
It works when I do not use var_list.

my code looks like this:
```
import tensorflow as tf
import tensorlayer as tl
import numpy as np
import random


x_train, y_train, x_val, y_val, x_test, y_test = tl.files.load_mnist_dataset(shape=(-1, 28, 28, 1))

# sample = x_train[2]
# sample = np.reshape(sample, [28, 28])
# im = plt.imshow(sample)
# plt.show()


x_train = np.asarray(x_train, np.float32)
y_train = np.asarray(y_train, np.float32)
x_val = np.asarray(x_val, np.float32)
y_val = np.asarray(y_val, np.float32)
x_test = np.asarray(x_test, np.float32)
y_test = np.asarray(y_test, np.float32)

x_train = np.concatenate([x_train, x_val])
y_train = np.concatenate([y_train, y_val])

g_input = tf.placeholder(tf.float32, [None, 100], name='noise')
d_input = tf.placeholder(tf.float32, [None, 28, 28, 1], name='mnist')
d_output = tf.placeholder(tf.float32, [None, 2], name='d_output')

W_init = tf.truncated_normal_initializer(stddev=0.01)
b_init = tf.constant_initializer(value=0.0)
relu = tf.nn.relu

batch_size = 128
lr = 1e-3

sess = tf.InteractiveSession()

# generative network
def G(reuse=False):
    chs = 128
    with tf.variable_scope('generator', reuse=reuse):
        tl.layers.set_name_reuse(reuse)
        g = tl.layers.InputLayer(g_input, name='g_input')
        g = tl.layers.DenseLayer(g, chs*14*14, name='fc1')
        g = tl.layers.ReshapeLayer(g, [-1, 14, 14, chs])
        g = tl.layers.UpSampling2dLayer(g, size=[2, 2], name='upsampling1')
        g = tl.layers.Conv2d(g, 54, (5, 5), padding='SAME', W_init=W_init, b_init=b_init, name='conv1')
        g = tl.layers.BatchNormLayer(g, act=relu, name='bn1')
        g = tl.layers.Conv2d(g, 32, (3, 3), padding='SAME', W_init=W_init, b_init=b_init, name='conv2')
        g = tl.layers.BatchNormLayer(g, act=relu, name='bn2')
        g = tl.layers.Conv2d(g, 1, (3, 3), padding='SAME', W_init=W_init, b_init=b_init, name='g_output')
        return g

g = G(reuse=False)

# discriminator
def D(reuse=False):
    with tf.variable_scope('discriminitor', reuse=reuse):
        tl.layers.set_name_reuse(reuse)
        d = tl.layers.InputLayer(d_input, name='d_input')
        d = tl.layers.Conv2d(d, 32, (3, 3), padding='SAME', W_init=W_init, b_init=b_init, name='d_conv1')
        d = tl.layers.BatchNormLayer(d, act=relu, name='d_bn1')
        d = tl.layers.MaxPool2d(d, (2, 2))
        d = tl.layers.Conv2d(d, 64, (3, 3), padding='SAME', W_init=W_init, b_init=b_init, name='d_conv2')
        d = tl.layers.BatchNormLayer(d, act=relu)
        d = tl.layers.MaxPool2d(d, (2, 2), name='d_bn2')
        d = tl.layers.FlattenLayer(d, name='d_flatten')
        # d = tl.layers.DropoutLayer(d, keep=0.5, name='dropout1')
        d = tl.layers.DenseLayer(d, 512, act=relu, name='fc2')
        # d = tl.layers.DropoutLayer(d, keep=0.5, name='dropout2')
        d = tl.layers.DenseLayer(d, 2, name='fc3')
        return d

d = D(reuse=False)
print 'params:'
print d.all_params



# objective function
loss_g = tf.reduce_mean(tf.losses.softmax_cross_entropy(d_output, d.outputs))
loss_d = tf.reduce_mean(tf.losses.softmax_cross_entropy(d_output, d.outputs))

# train op
train_g = tf.train.AdamOptimizer(lr).minimize(loss_g, var_list=g.all_params)
train_d = tf.train.AdamOptimizer(lr).minimize(loss_d, var_list=d.all_params)


# pre-train for discriminator
sess.run(tf.global_variables_initializer())
for i in xrange(100):
    feature1 = np.random.uniform(0, 1, size=(batch_size/2, 28, 28, 1))
    feature2 = x_train[random.sample(range(0, x_train.shape[0]), batch_size/2)]
    feature = np.concatenate([feature1, feature2])
    label = np.zeros((batch_size, 2))
    for j in xrange(batch_size):
        if j < batch_size/2:
            label[j, 1] = 1.0
        else:
            label[j, 0] = 1.0
    loss_val, _ = sess.run([loss_d, train_d], feed_dict={d_input: feature, d_output: label, g_input: np.random.uniform(0, 1, size=(batch_size, 100))})
    # tl.utils.fit(sess, d, train_d, loss_d, x_train, y_train, d_input, d_output,batch_size=batch_size, n_epoch=500, print_freq=5, eval_train=False)
    print 'at step %d:, loss value is %f'%(i+1, loss_val)


sess.close()

```

**Comment by zsdonghao** [comment]
@fangpin pls follow this implementation https://github.com/zsdonghao/dcgan

**Comment by fangpin** [comment]
@zsdonghao thanks for your nice work
